### ISSUE TITLE
No wildcard redirects possible?

### ISSUE THREAD

**Issue by marcelstoer** [rtfd/readthedocs.org#3465]
When we started with RTD & MkDocs two years ago we tried to work around some l10n shortcomings ourselves. Turns out we don't need that all; all files can be moved up one level in the directory tree. To avoid dead links we'd like to configure wild card redirects to omit that extra folder in the URL.

Examples:
- https://nodemcu.readthedocs.io/en/latest/en/ -> https://nodemcu.readthedocs.io/en/latest/
- https://nodemcu.readthedocs.io/en/latest/en/flash/ -> https://nodemcu.readthedocs.io/en/latest/flash/
- https://nodemcu.readthedocs.io/en/latest/en/modules/cron/ -> https://nodemcu.readthedocs.io/en/latest/modules/cron/

In short, in mod_rewrite terms this would be `/$lang/$version/en/(.*) -> /$lang/$version/$1`.

I studied the three redirect options in the documentation and ran a few tests but a) I'm either too stupid to get it or b) then such wild card redirects really aren't supported. If b) then https://github.com/honzajavorek/rtd-redirects might come in handy for us.

**Comment by marcelstoer** [comment]
The 'Enhancement' tag added by @RichardLitt suggests that it's b) indeed...what a relief 😜

**Comment by RichardLitt** [comment]
Best guess on my part. Let's see what the others have to say. :)

**Comment by vidartf** [comment]
I need to do something similar, so I looked at code here: https://github.com/rtfd/readthedocs.org/blob/96f6da57d89a37bdfa6b99de93e8ae6d09e2a586/readthedocs/redirects/models.py

There does not seem to be any wildcards, but I see that exact redirects have a '$rest' parameter. I think this will allow you to do what you want, by the following pattern:
 - from: `/$lang/$version/en/$rest`
 - to: `/$lang/$version/`
(you might need to tweak the start, not sure how from/to gets set before entering code linked).

Note: The `$rest` bit seems not to be documented, so here be dragons.

**Comment by marcelstoer** [comment]
You mean https://github.com/rtfd/readthedocs.org/blob/96f6da57d89a37bdfa6b99de93e8ae6d09e2a586/readthedocs/redirects/models.py#L134 right? I don't see support for either `$lang` or `$version` in the code though.

I was recently testing redirects _w/ modifying my file structure first_ (i.e. I didn't _have_ any invalid links) and was then wondering why my redirect rules weren't applied at all. Turns out [they're only active after HTTP 404 was detected](https://docs.readthedocs.io/en/latest/user-defined-redirects.html#implementation):
> Since we serve documentation in a highly available way, we do not run any logic when we’re serving documentation. This means that redirects will only happen in the case of a _404 File Not Found_.

**Comment by vidartf** [comment]
> I don't see support for either `$lang` or `$version` in the code though.

Yes, looking closer on this I see that. Are you sure your case can't be covered by the prefix redirect?

**Comment by fantix** [comment]
I have a case needing wildcard redirects that one language of translation is completely hosted somewhere else, so I'd like to redirect `/en/(.*)` to `https://else.where/en/$1`.

**Comment by stsewd** [comment]
The rest keyword is now documented https://docs.readthedocs.io/en/latest/user-defined-redirects.html#exact-redirects Looks like it solves all the problems here, not sure if a wildcard is still needed after that.

**Comment by stsewd** [comment]
I believe `/en/latest/en/$rest` -> `/en/latest/` should work (I just tested this and it works!). I'm not completely sure, but for me looks like the rest keyword is the _wildcard redirect_  p:

**Comment by humitos** [comment]
@marcelstoer could you please test the solution that @stsewd provided? I think that should achieve your goal.

I want to have something actionable on this issue or close it.

**Comment by marcelstoer** [comment]
@humitos it's great that you documented this behavior back in August in https://github.com/rtfd/readthedocs.org/commit/f4348591a42ebcc8a34166c4ee260e90897868c8. What I didn't get initially is that `$rest` doesn't need to be repeated in the to-URL (thinking about regex capturing groups and replacements). Triggered by the bot message above I ran a few tests in my project and all is well. Thanks again for your contributions and support.
